### DISCUSSION TITLE
Database Encryption

### DISCUSSION THREAD

**Discussion by mhmadalaa** [lilhind/passwordManager#19] [Show and tell]
### Let's answer these questions to show what's going on

- How client send passowrd?
- How password stroed in databse?
- How password retrieved to the client?
- How Encryption actually works?

## How client send passowrd?
```json
{
  "hubName": "twitter",
  "email": "fff@gggz.com",
  "password": "loloaaa 33444"
}
```

## How password stroed in databse?
```json
{
  "_id": {
    "$oid": "653c0b6c74ab09b519542fed"
  },
  "hubName": "twitter",
  "email": "fff@gggz.com",
  "password": "af11bc2dc137604b479f0b1503964265",
  "cipheriv": {
    "$binary": {
      "base64": "OFIoARtcVhwcJ8+HkC7x5Q==",
      "subType": "00"
    }
  },
  "passwordChangedAt": {
    "$date": "2023-10-27T19:11:40.985Z"
  },
  "__v": 0
}
```

## How password retrieved to the client?
```json
{
  "_id": "653c0b6c74ab09b519542fed",
  "hubName": "twitter",
  "email": "fff@gggz.com",
  "passwordChangedAt": "2023-10-27T19:11:40.985Z",
  "hubPassword": "loloaaa 33444"
}
```

## How Encryption actually works?
We use `crypto` a nodejs module, it needs cipheriv and key to encrypt and decrypt
- For cipheriv we made it unique for each password, and actually stored with it 
  - ```this.cipheriv = crypto.randomBytes(16);```
- For key we do it as an environment variable random bytes encoded in base64 
  - ``` const key = Buffer.from(process.env.CIPHER_SECRET_KEY, 'hex').toString('base64',); ```

**Comment by ahmedhindawi** [comment]


**Reply by mhmadalaa** [reply]
بنتعلم منك
![image](https://github.com/lilhind/passwordManager/assets/68995755/1fc4c0ee-71ca-416c-b44c-9539c21d2ee5)
